Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/event-native-apim/event-native-apim-example-use-cases.adoc

[[event-native-apim-example-use-cases]]
= Event-native API Management example use cases
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_example_use_cases.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

[NOTE]
====
This document assumes you are familiar with synchronous APIs, asynchronous APIs, and the OpenAPI specification.
====

This page describes some example use cases for event-native API management.

== Prerequisites

=== Kafka

These examples all depend on Kafka. To set up a Kafka Docker container with the link:https://github.com/tchiotludo/akhq[AKHQ] UI, follow these steps.

1. Download `link:https://raw.githubusercontent.com/tchiotludo/akhq/master/docker-compose.yml[docker-compose.yml]`.

2. Run it locally.
+
[source, bash]
----
docker-compose pull
docker-compose up
----

You can now access the AKHQ UI at http://localhost:8080.

===  Enabling event-native API management

> To run these examples, you must set the `gravitee_api_jupiterMode_enabled` environment variable to `true` on the Management API and the API Gateway.

Reviewer: What is expected is unclear. How this must be done?

Author: I have added a couple of lines explaining a bit more.

=== Postman Collections

These examples use the link:https://github.com/gravitee-io/postman-collections[Gravitee Postman Collections].

These collections use four variables.

[cols="1,1,1", options="header"]
|===
| Variable 
| Description 
| Example

| `management_host`
| The host for the Management API
| `http://localhost:8083` 

| `management_username`
| The username for a management user
| `admin`

| `management_password`
| The password for `management_username`
| 

| `gateway_host`
| The gateway host
| `http://loalhost:8082`
|===

> == Data ingestion

Reviewer: Some explanation could be great on the different use case achievable thank to api v4.

Author: Yes, I will add this in the next iteration.

Author: I have added a diagram showing some more detail here.

For data ingestion, run the _01 - Data Ingestion_ Postman Collection.

You can also use `curl` to `POST` data to the endpoint.

[source bash]
----
curl -X POST -d "my_payload" http://localhost:8082/data/ingestion
----

== Event consumption

=== Streaming: server-sent events (SSE)

For streaming with server-sent events (SSE), run the _02 - Event Consumption - SSE_ Postman Collection.

You can test it with `curl`.

[source bash]
----
curl -N -H "Accept:text/event-stream" http://localhost:8082/demo/sse
----

=== Streaming: WebSocket

For streaming with WebSocket, run the _03 - Event Consumption - Websocket_ Postman Collection.

> You can test it with Postman's WebSocket support, or you can use the `websocat` command-line tool.

Reviewer: How ?

[source bash]
----
websocat ws://localhost:8082/demo/ws
----

=== Webhooks

For webhooks, run the _04 - Event Consumption - Webhook_ Postman Collection.

This collection uses a webhook callback that is called by the API Gateway. The collection uses https://webhook.site/.

To use this collection, go to https://webhook.site/ to get your unique callback URL, and update the Postman Collection to use it. For example:

[source json]
----
{
    "configuration": {
        "type": "webhook",
        "callbackUrl": "https://webhook.site/891490b9-1e37-4b5e-8f91-4d40b9187710"
    }
}
----

==== Webhooks with subscription filter

For webhooks with subscription filters, run the _05 - Event Consumption - Webhook - Message Filtering_ Postman Collection.

Use the following policy configuration.

[source json]
----
{
  "name": "Message filtering",
  "description": "Apply filter to messages",
  "enabled": true,
  "policy": "message-filtering",
  "configuration": {

>     "filter": "{#subscription.filter}"

Reviewer: Wrong configuration here.

Author: What is the correct configuration?

Reviewer 2: ```
 {
                    "name": "Message filtering",
                    "description": "Apply filter to messages",
                    "enabled": true,
                    "policy": "message-filtering",
                    "configuration": {
                        "filter": "{#jsonPath(#message.content, '$.feature') == #subscription.metadata.feature}"
                    }
                }
```

[NOTE]
====
This document assumes you are familiar with synchronous APIs, asynchronous APIs, and the OpenAPI specification.
====

This page describes some example use cases for event-native API management.

== Prerequisites

=== Kafka

These examples all depend on Kafka. To set up a Kafka Docker container with the link:https://github.com/tchiotludo/akhq[AKHQ] UI, follow these steps.

1. Download `link:https://raw.githubusercontent.com/tchiotludo/akhq/master/docker-compose.yml[docker-compose.yml]`.

2. Run it locally.
+
[source, bash]
----
docker-compose pull
docker-compose up
----

You can now access the AKHQ UI at http://localhost:8080.

===  Enabling event-native API management

To run these examples, you must set the `gravitee_api_jupiterMode_enabled` environment variable to `true` on the Management API and the API Gateway.

See link:{{'/apim/3.x/apim_installguide_rest_apis_configuration.html#environment_variables' | relative_url }}[Environment variables] on the Management API and link:{{'/apim/3.x/apim_installguide_gateway_configuration.html#environment_variables' | relative_url }}[Environment variables] on the API Gateway for more detail about setting environment variables.

> === Postman Collections

Reviewer 3: ```suggestion
=== V4 BETA 3.20.x Postman Collection
```

> These examples use the link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Postman Collections].

Reviewer 3: ```suggestion
These examples use the Gravitee V4 BETA 3.20.x Postman Collection, available from the link:https://www.postman.com/gravitee-io/workspace/gravitee-public-workspace/overview[Gravitee Public Workspace] in Postman.
```

> These collections use four variables.

Reviewer 3: ```suggestion
This collection uses four variables.
```

[cols="1,1,1", options="header"]
|===
| Variable 
| Description 
| Example

| `management_host`
| The host for the Management API
| `http://localhost:8083` 

| `management_username`
| The username for a management user
| `admin`

| `management_password`
| The password for `management_username`
| 

| `gateway_host`
| The gateway host
| `http://loalhost:8082`
|===

== Data ingestion

Some data ingestion examples are illustrated in the following diagram.

image:{% link /images/apim/3.x/event-native/event-native-api-management-data-ingestion.png %}[]

> For data ingestion, run the _01 - Data Ingestion_ Postman Collection.

Reviewer 3: ```suggestion
For data ingestion, run the requests in the _01 - Data Ingestion_ folder of the Postman Collection.
```

> You can also use `curl` to `POST` data to the endpoint.

Reviewer 3: ```suggestion
You can also use `curl` to `POST` data to the endpoint, as shown in the example below.
```

[source bash]
----
curl -X POST -d "my_payload" http://localhost:8082/data/ingestion
----

== Event consumption

=== Streaming: server-sent events (SSE)

> For streaming with server-sent events (SSE), run the _02 - Event Consumption - SSE_ Postman Collection.

Reviewer 3: ```suggestion
For streaming with server-sent events (SSE), run the requests in the _02 - Event Consumption - SSE_ folder of the Postman Collection.
```

You can test it with `curl`.

[source bash]
----
curl -N -H "Accept:text/event-stream" http://localhost:8082/demo/sse
----

=== Streaming: WebSocket

> For streaming with WebSocket, run the _03 - Event Consumption - Websocket_ Postman Collection.

Reviewer 3: ```suggestion
For streaming with WebSocket, run the requests in the _03 - Event Consumption - Websocket_ folder of the Postman Collection.
```

You can test it through a WebSocket connection in Postman, or you can use the `websocat` command-line tool as shown in the example below.

[source bash]
----
websocat ws://localhost:8082/demo/ws
----

=== Webhooks

> For webhooks, run the _04 - Event Consumption - Webhook_ Postman Collection.

Reviewer 3: ```suggestion
For webhooks, run the requests in the _04 - Event Consumption - Webhook_ folder of the Postman Collection.
```

> This collection uses a webhook callback that is called by the API Gateway. The collection uses https://webhook.site/.

Reviewer 3: ```suggestion
This request group uses a webhook callback that is called by the API Gateway. The unique callback URL is generated via https://webhook.site/.
```

> To use this collection, go to https://webhook.site/ to get your unique callback URL, and update the Postman Collection to use it. For example:

Reviewer 3: ```suggestion
To use these requests, go to https://webhook.site/ to get your unique callback URL, and update the Postman Collection to use it. For example:
```

[source json]
----
{
    "configuration": {
        "type": "webhook",
        "callbackUrl": "https://webhook.site/891490b9-1e37-4b5e-8f91-4d40b9187710"
    }
}
----

==== Webhooks with subscription filter

> For webhooks with subscription filters, run the _05 - Event Consumption - Webhook - Message Filtering_ Postman Collection.

Reviewer 3: ```suggestion
For webhooks with subscription filters, run the requests in the _05 - Event Consumption - Webhook - Message Filtering_ folder of the Postman Collection.
```
